src/constants.ts

```diff
@@ -54,3 +54,50 @@ export const renderTypeOptions: Record<Exclude<QuestionType, 'obs'>, Array<Rende
   patientIdentifier: ['text'],
   programState: ['select'],
 };
+
+/**
+  Required properties for all question types.
+ */
+export const requiredProperties: Array<keyof FormField> = ['id', 'label', 'type', 'questionOptions'];
+
+/**
+   Type-specific properties for each question type.
+ */
+export const typeSpecificProperties: Record<string, Array<keyof FormField>> = {
+  control: [],
+  encounterDatetime: ['datePickerFormat'],
+  encounterLocation: [],
+  encounterProvider: [],
+  encounterRole: [],
+  obs: ['required'],
+  obsGroup: ['questions'],
+  patientIdentifier: [],
+  testOrder: [],
+  programState: [],
+};
+
+/**
+  Merge required properties with type-specific ones.
+ */
+export const allowedPropertiesMapping: Record<string, string[]> = Object.fromEntries(
+  Object.entries(typeSpecificProperties).map(([type, props]) => {
+    const mergedProps = new Set<string>([...requiredProperties, ...props]);
+    return [type, Array.from(mergedProps)];
+  }),
+);
+
+/**
+   Mapping of allowed keys for the nested questionOptions object per question type.
+ */
+export const allowedQuestionOptionsMapping: Record<string, string[]> = {
+  control: ['rendering', 'minLength', 'maxLength'],
```

Comment:
'minLength', 'maxLength' are properties relevant to `text` rendering type and not specific to the `control` question type